proto/cacheclient.proto

```diff
@@ -402,16 +402,110 @@ message _ListRemoveRequest {
 
 message _ListRemoveResponse {}
 
+/*
+  # ListFetch and ListRetain documentation and examples
+
+  # start and end are zero-based indexes, both are optional
+
+  # start is inclusive
+  # If start is not set, then the default is Unbounded, which is treated as 0
+  # Unbounded == 0,
+  # this field has a explicit presence for future compatibility to support reverse order and controlling our stride
+
+  # end is exclusive
+  # If end is not set, then the default is Unbounded which becomes (list.len)
+
+  # start and end can be negative numbers, negative indicates offsets starting at the end of the list.
+  # For example, -1 is the last element of the list, -2 the penultimate, etc
+  #               [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
+  # +start/+end    0, 1, 2, 3, 4, 5, 6, 7, 8, 9
+  # -start/-end  -10,-9,-8,-7,-6,-5,-4,-3,-2,-1
+
+  list_ = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
+
+  # Using the slice notation:
+  # start:end
+
+  # start=0, end=Unbounded
+  list_[0:] == [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
+
+  # start=5, end=Unbounded
+  list_[5:] == [5, 6, 7, 8, 9]
+
+  # start=5, end=0
+  list_[5:0] == []
+
+  # start=-5, end=Unbounded
+  list_[-5:] == [5, 6, 7, 8, 9]
+
+  # start=-5, end=0
+  list_[-5:0] == []
+
+  # start=0, end=5
+  list_[0:5] == [0, 1, 2, 3]
+
+  # start=Unbounded, end=5
+  list_[0:5] == [0, 1, 2, 3]
+
+  # start=0, end=-5
+  list_[0:-5] == [0, 1, 2, 3, 4]
+
+  # start=Unbounded, end=-5
+  list_[0:-5] == [0, 1, 2, 3, 4]
+
+  # start=2, end=6
+  list_[2:6] == [2, 3, 4, 5]
+
+  # start=2, end=-6
+  list_[2:-6] == [2, 3]
+
+  # start=-7, end=-2
+  list_[-7:-2] == [3, 4, 5, 6, 7]
+
+  // if using ltrim, this will remove all elements in the list
+  # start=-2, end=-6
+  list_[2:-6] == []
```

Comment:
Why should we interpret an empty range that _looks_ like a partial deletion was _intended_ in this 1 special case as "delete your entire list" instead of the more conservative "do nothing?"
Isn't this inconsistent with the fetch behavior, where that same partial range intent doesn't end up mean "return me the whole list?

This is a footgun, we should talk about it before moving ahead.

Reply:
I agree, I think if you bounds are invalid, then we shouldn't delete the entire list. (also, the comment says start=-1, end=-6, but notation says [2:-6], notation here is wrong.) 

Previously, we styling after redis implementation of of ltrim, invalid bounds would cause the list to be removed/emptied, I would lean more towards returning a invalid bounds response, how do we feel about

```
message _ListRetainResponse {
  message _Success {}

  message _InvalidBounds {}

  oneof list {
    _Success success = 1;
    _InvalidBounds invalid = 2;
  }
}
```?

Reply:
Oh interesting, I don't have a strong opinion about whether to prefer [] or _InvalidBounds. It would probably be nicer for ListRetain to let you know you tried to do something that can't ever do anything. But then I wonder if it should be done for fetch as well?

Reply:
But, to clarify, this would only be for invalid bounds, wouldn't work for bounds that are valid, but out of range of your list. You can try and retain out of bounds, but that would mean trimming the entire list.           
```
list - [0,1,2,3,4,5,6,7,8,9]
ListRetain(start_index: 100, end_index: 102)
list - []
```

Yeah, not opposed to adding to ListFetch

Reply:
I can see a case for an error for `ListRetain`, ie if they want to delete the whole thing they should do that. But for `ListFetch` I prefer returning an empty slice as that is a convention familiar in various languages.

Should we go with a gRPC error, I prefer `OUT_OF_RANGE` (see [here](https://grpc.github.io/grpc/core/md_doc_statuscodes.html)) as that maps cleanly to our use case and is otherwise unambiguous. For other list indexing operations (eg "get the value at position 10") this is appropriate as well.